File: packages/core/tests/card-layers.test.ts
import { describe, it, expect } from 'vitest';
import {
  DEPTH_LAYERS_KEY,
  LAYER_DEPTH_ORDER,
  OPTIONS_LAYER_DEPTH,
  depthToParallaxFactor,
  getCardDepthLayers,
  validateCardDepthLayers,
} from '../src/card-layers';
import type { CardDepthLayer } from '../src/types';

describe('KRO-130 — depthToParallaxFactor', () => {
  it('crece de fondo a frente (back < mid < front)', () => {
    expect(depthToParallaxFactor('back')).toBeLessThan(depthToParallaxFactor('mid'));
    expect(depthToParallaxFactor('mid')).toBeLessThan(depthToParallaxFactor('front'));
  });

  it('todos los factores están en 0..1', () => {
    for (const d of LAYER_DEPTH_ORDER) {
      const f = depthToParallaxFactor(d);
      expect(f).toBeGreaterThanOrEqual(0);
      expect(f).toBeLessThanOrEqual(1);
    }
  });

  it('valor fuera de catálogo → 0.5 neutro', () => {
    // @ts-expect-error — probamos robustez ante un depth no válido en runtime
    expect(depthToParallaxFactor('whatever')).toBe(0.5);
  });
});

describe('KRO-130 — OPTIONS_LAYER_DEPTH', () => {
  it('tiene los 3 presets y sus ids casan con LAYER_DEPTH_ORDER', () => {
    expect(OPTIONS_LAYER_DEPTH.map(o => o.id)).toEqual([...LAYER_DEPTH_ORDER]);
    for (const o of OPTIONS_LAYER_DEPTH) expect(o.label).toBeTruthy();
  });
});

describe('KRO-130 — getCardDepthLayers', () => {
  const mk = (depth: string, url = 'x.png'): CardDepthLayer => ({ depth: depth as any, url });

  it('carta sin capas → []', () => {
    expect(getCardDepthLayers(undefined)).toEqual([]);
    expect(getCardDepthLayers({})).toEqual([]);
    expect(getCardDepthLayers({ name: 'C1' })).toEqual([]);
  });

  it('lee del campo reservado y ordena fondo→frente', () => {
    const card = { [DEPTH_LAYERS_KEY]: [mk('front', 'p.png'), mk('back', 'bg.png'), mk('mid', 's.png')] };
    expect(getCardDepthLayers(card).map(l => l.depth)).toEqual(['back', 'mid', 'front']);
  });

  it('descarta entradas malformadas (sin url o depth inválido)', () => {
    const card = {
      [DEPTH_LAYERS_KEY]: [
        mk('back', 'bg.png'),
        { depth: 'mid' },                 // sin url
        { url: 'x.png', depth: 'nope' },  // depth inválido
        null,
      ],
    };
    expect(getCardDepthLayers(card)).toHaveLength(1);
    expect(getCardDepthLayers(card)[0].url).toBe('bg.png');
  });

  it('campo presente pero no-array → []', () => {
    expect(getCardDepthLayers({ [DEPTH_LAYERS_KEY]: 'oops' })).toEqual([]);
  });
});

describe('KRO-130 — validateCardDepthLayers', () => {
  it('lista vacía o ausente → válida', () => {
    expect(validateCardDepthLayers(undefined).valid).toBe(true);
    expect(validateCardDepthLayers([]).valid).toBe(true);
  });

  it('capas correctas (con y sin foil) → válida', () => {
    const layers: CardDepthLayer[] = [
      { url: 'bg.png', depth: 'back' },
      { url: 's.png', depth: 'mid', foil: { kind: 'foil', textureUrl: 'foil.png', blend: 'screen' } },
    ];
    expect(validateCardDepthLayers(layers).valid).toBe(true);
  });

  it('reporta url vacía, depth inválido y foil sin textura', () => {
    const layers = [
      { url: '', depth: 'back' },
      { url: 'x.png', depth: 'wrong' },
      { url: 'y.png', depth: 'front', foil: { kind: 'foil', textureUrl: '', blend: 'screen' } },
    ] as unknown as CardDepthLayer[];
    const r = validateCardDepthLayers(layers);
    expect(r.valid).toBe(false);
    expect(r.issues.map(i => i.index).sort()).toEqual([0, 1, 2]);
  });
});

File: packages/core_dart/test/card_layers_test.dart
import 'package:kromia_core/kromia_core.dart';
import 'package:test/test.dart';

/// KRO-130/122 — espejo Dart de `card-layers.ts` (capas de profundidad 3D +
/// parallax + foil). Valores verificados 1:1 contra el source TS.
void main() {
  // Las constantes canónicas (parallax, blend modes, orden, claves) viven en el
  // drift-test cruzado: `card_layers_drift_test.dart`.

  group('getCardDepthLayers', () {
    test('null / sin clave / no-lista → []', () {
      expect(getCardDepthLayers(null), isEmpty);
      expect(getCardDepthLayers({}), isEmpty);
      expect(getCardDepthLayers({depthLayersKey: 'nope'}), isEmpty);
    });

    test('descarta malformadas (sin url / depth inválida) y reordena back→mid→front', () {
      final card = {
        depthLayersKey: [
          {'url': 'f.png', 'depth': 'front'},
          {'depth': 'back'}, // sin url → descartada
          {'url': 'b.png', 'depth': 'back'},
          {'url': 'x.png', 'depth': 'side'}, // depth inválida → descartada
          {'url': 'm.png', 'depth': 'mid'},
        ],
      };
      final layers = getCardDepthLayers(card);
      expect(layers.map((l) => l.depth).toList(), ['back', 'mid', 'front']);
      expect(layers.map((l) => l.url).toList(), ['b.png', 'm.png', 'f.png']);
    });

    test('parsea el foil vinculado a una capa', () {
      final card = {
        depthLayersKey: [
          {
            'url': 's.png',
            'depth': 'mid',
            'foil': {
              'kind': 'foil',
              'textureUrl': 'foil.png',
              'maskUrl': 'mask.png',
              'blend': 'color-dodge',
              'intensity': 0.6,
            },
          },
        ],
      };
      final foil = getCardDepthLayers(card).single.foil!;
      expect(foil.kind, 'foil');
      expect(foil.textureUrl, 'foil.png');
      expect(foil.maskUrl, 'mask.png');
      expect(foil.blend, 'color-dodge');
      expect(foil.intensity, 0.6);
    });
  });

  group('validateCardDepthLayers', () {
    test('null / vacía → válida (es opt-in)', () {
      expect(validateCardDepthLayers(null).valid, isTrue);
      expect(validateCardDepthLayers(const []).valid, isTrue);
    });

    test('url vacía, depth inválida y foil sin textura → issues', () {
      final res = validateCardDepthLayers([
        const CardDepthLayer(url: '', depth: 'back'),
        const CardDepthLayer(url: 'ok.png', depth: 'nope'),
        const CardDepthLayer(
          url: 'ok.png',
          depth: 'mid',
          foil: EffectLayer(kind: 'foil', textureUrl: '', blend: 'overlay'),
        ),
      ]);
      expect(res.valid, isFalse);
      expect(res.issues.length, 3);
      expect(res.issues.map((i) => i.index).toSet(), {0, 1, 2});
    });

    test('capa válida con foil correcto → sin issues', () {
      final res = validateCardDepthLayers([
        const CardDepthLayer(
          url: 's.png',
          depth: 'mid',
          foil: EffectLayer(kind: 'foil', textureUrl: 't.png', blend: 'screen'),
        ),
      ]);
      expect(res.valid, isTrue);
      expect(res.issues, isEmpty);
    });
  });

  group('EffectLayer.fromJson', () {
    test('lee kind/textureUrl/maskUrl/blend/intensity/motion', () {
      final e = EffectLayer.fromJson(const {
        'kind': 'glitter',
        'textureUrl': 't.png',
        'maskUrl': 'm.png',
        'blend': 'soft-light',
        'intensity': 0.4,
        'motion': 0.8,
      });
      expect(e.kind, 'glitter');
      expect(e.textureUrl, 't.png');
      expect(e.maskUrl, 'm.png');
      expect(e.blend, 'soft-light');
      expect(e.intensity, 0.4);
      expect(e.motion, 0.8);
    });
  });
}
